group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -144,216 +133,137 @@ protected GroupAssignment buildAssignment() throws PartitionAssignorException {
-        Map<Uuid, Set<Integer>> currentMemberAssignment
-    ) {
-        List<TopicIdPartition> validCurrentAssignmentList = new ArrayList<>();
-        currentMemberAssignment.forEach((topicId, partitions) -> {
-            if (subscribedTopicIds.contains(topicId)) {
-                partitions.forEach(partition -> {
-                    TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, partition);
-                    validCurrentAssignmentList.add(topicIdPartition);
-                });
+            if (newAssignment == null) {
+                targetAssignment.put(memberId, new MemberAssignment(oldAssignment));
             } else {
-                LOG.debug("The topic " + topicId + " is no longer present in the subscribed topics list");
+                targetAssignment.put(memberId, new MemberAssignment(newAssignment));
             }
-        });
-
-        return validCurrentAssignmentList;
+        }
     }
 
-    /**
-     * Allocates the unassigned partitions to unfilled members in a round-robin fashion.
-     */
-    private void unassignedPartitionsRoundRobinAssignment() {
-        Queue<String> roundRobinMembers = new LinkedList<>(potentiallyUnfilledMembers.keySet());
-
-        // Partitions are sorted to ensure an even topic wise distribution across members.
-        // This not only balances the load but also makes partition-to-member mapping more predictable.
-        List<TopicIdPartition> sortedPartitionsList = unassignedPartitions.stream()
-            .sorted(Comparator.comparing(TopicIdPartition::topicId).thenComparing(TopicIdPartition::partitionId))
-            .collect(Collectors.toList());
-
-        for (TopicIdPartition topicIdPartition : sortedPartitionsList) {
-            boolean assigned = false;
-
-            for (int i = 0; i < roundRobinMembers.size() && !assigned; i++) {
-                String memberId = roundRobinMembers.poll();
-                if (potentiallyUnfilledMembers.containsKey(memberId)) {
-                    assigned = maybeAssignPartitionToMember(memberId, topicIdPartition);
-                }
-                // Only re-add the member to the end of the queue if it's still available for assignment.
-                if (potentiallyUnfilledMembers.containsKey(memberId)) {
-                    roundRobinMembers.add(memberId);
-                }
+    private void assignRemainingPartitions() {
+        int unassignedPartitionIndex = 0;
+
+        for (MemberWithRemainingQuota unfilledMember : potentiallyUnfilledMembers) {
+            String memberId = unfilledMember.memberId;
+            int remainingQuota = unfilledMember.remainingQuota;
+
+            Map<Uuid, Set<Integer>> newAssignment = targetAssignment.get(memberId).targetPartitions();
+            if (isImmutableMap(newAssignment)) {
+                newAssignment = deepCopy(newAssignment);
+                targetAssignment.put(memberId, new MemberAssignment(newAssignment));
```

Comment:
i think what we're doing here is that for all unfilled members, since we have to modify their assignments we are essentially checking if they were not modified during revoke stage and if not to transform the map to a mutable one. is this correct?

i think it will be worth to add a comment here

Reply:
Yeah, just so I understand, we will copy at most once if something is revoked or added, but if nothing changes, there is no copy?

Reply:
@jeffkbkim Correct. Added a comment.

@jolshan Correct too. We basically copy only if we want to alter the assignment.